[Community] Add fairseq FastSpeech2

jaketae opened the pull request:
# What does this PR do?

This PR adds FastSpeech2, a transformer-based text-to-speech model. Fixes #15166.

## Motivation

While HF transformers has great support for ASR and other audio processing tasks, there is not much support for text-to-speech. There are many transformer-based TTS models that would be great to have in the library. FastSpeech2 is perhaps the most well-known and popular of transformer-based TTS models. With TTS support, we could potentially have really cool demos that go from ASR -> NLU -> NLG -> TTS to build full-fledged voice assistants! 

## Who can review?

Anton, Patrick (will tag when PR is ready)

Commit:
wip: add fairseq fastspeech2

Commit:
fix: use bart attn, complete ckpt conversion

Commit:
style: run linter

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15773). All of your documentation changes will be reflected on that endpoint.

Commit:
chore: rm non existent classes

Commit:
chore: rm unused tf script, label cp origin

Commit:
chore: cleanup dead code

Commit:
feature: add mean std for mel normalization

Commit:
wip: write tokenizer code

Commit:
chore: rm fast tokenizer

Commit:
chore: rm fast tokenizer

Commit:
feat: add g2p_en backend, complete tokenizer

Commit:
fix: add padding token

Commit:
style: run linter

Commit:
fix: stash g2p attr, add eos, rm unused import

jaketae commented:
Here is a quick update on this PR, summarizing some of the things discussed with @anton-l for public visibility.

1. Checkpoint loading: The model weight loads properly and produces roughly the same output as the fairseq model. On preliminary tests, I've found that `torch.allclose` pass with `atol=1e-6`. On this branch, the model can be loaded via
    ```python
    from transformers import AutoModel
    model = AutoModel.from_pretrained("jaketae/fastspeech2-lj")
    ```
2. Grapheme-to-phoneme: Fairseq uses [`g2p_en`](https://github.com/Kyubyong/g2p). I've added this as a dependency via `requires_backend` in the tokenizer. One thing I'm not too sure about is whether to enforce this from the get-go at `__init__`, or only require it in the `phonemize` function. 
3. Tokenizer: The tokenizer can be loaded as shown. For some reason, `AutoTokenizer` doesn't work at the moment, but this can likely be fixed by reviewing import statements. One consideration is that the fairseq code seems to add an EOS token at the end of every sequence, whereas HF's tokenizers do not by default (correct me if I'm wrong). In my code, I manually add the EOS token at the end for consistency.
    ```python
    from transformers.models.fastspeech2 import FastSpeech2Tokenizer
    tokenizer = FastSpeech2Tokenizer.from_pretrained("jaketae/fastspeech2-lj")
    ```
4. Vocoder: (Note: This is not directly related to the source code in this PR.) Using Griffin-Lim for vocoding yields very poor audio quality and is likely a bad idea for demos. An alternative is to use neural vocoders like HiFi-GAN. I've added HiFi-GAN on the model hub, which can be imported via `vocoder = AutoModel.from_pretrained("jaketae/hifigan-lj-v1", trust_remote_code=True)`. At the moment, for some reason the model weights do not load properly, debugging required.

I have some questions for discussion.

1. Model output: What should the output of the model be? At the moment, it returns a tuple of mel-spectrograms, mel-spectrogram after postnet, output lengths (number of spectrogram frames), duration predictor output, predicted pitch, and predicted energy. I assume we need to subclass `BaseModelOutput` for something like `TTSModelOutput`.
2. TTS pipeline: At the moment, there is only `FastSpeech2Model`, which outputs mel-spectrograms as indicated above. While people might want to use acoustic feature generators for mel-spectrogram synthesis, more often they will want to obtain raw audio in waveform. As mentioned, this requires a vocoder, and we need to discuss ways to integrate this. Some API ideas I have include
- Introducing a `FastSpeech2ForAudioSynthesis` class with an integrated vocoder
- Adding a `pipeline` that supports full-fledged TTS. 

To-do's on my end (will be updated as I find more).
- [ ] Writing tests
- [ ] Documentation
- [ ] Cleaning up imports
- [ ] Debugging why `AutoTokenizer` fails
- [ ] Debugging why HiFi-GAN weights do not load

cc @patrickvonplaten

jaketae commented on src/transformers/models/fastspeech2/tokenization_fastspeech2.py:
```diff
@@ -0,0 +1,145 @@
+        preserve_punctuation=False,
+        **kwargs
+    ):
+        requires_backends(self, "g2p_en")
```
1. Should this be required here?
2. Do we want `do_phonemize` to be an option? It should probably always happen.

jaketae commented on src/transformers/models/fastspeech2/tokenization_fastspeech2.py:
```diff
@@ -0,0 +1,145 @@
+        tokens = list(filter(lambda p: p.strip() != "", tokens))
+
+        # fastspeech2 needs eos at the end
+        tokens.append(self.eos_token)
```
Adding EOS manually at the end of sequence.

jaketae commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,700 @@
+            )
+
+        if args.mean:
+            self.register_buffer("mean", torch.zeros(self.out_dim))
```
Mean and standard deviation are 0 and 1 by default so that even if the user forgets to use `set_mean` and `set_std`, the output is not affected.

patrickvonplaten replied:
Wouldn't it be cleaner to let `mean` be of type `float` and set it to `0.0` by default (same with variance)? The `if config.mean()` is a bit weird to me

jaketae replied:
I changed the flag name to `config.use_mean`. I also added an informative warning in case the user sets `use_mean` to `True`, pointing them to the `set_mean()` function (likewise for variance). 

Do you think it makes more sense for the user to specify the mean and variance explicitly in the config object? If so, one problem is that the mean and variance are 80-d vectors, so I'm afraid that it would pollute `config.json` too much. I might have misunderstood your intent though, definitely open to suggestions as I also think there might be a cleaner way to achieve this!

Commit:
wip: rm unused classes in test

anton-l commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,700 @@
+        processor_class=_TOKENIZER_FOR_DOC,
+        checkpoint=_CHECKPOINT_FOR_DOC,
+        # NOTE: should change model output type
+        output_type=BaseModelOutputWithPastAndCrossAttentions,
```
Feel free to put the `FastSpeech2BaseModelOutput()` dataclass right at the top of this file (similar to `Wav2Vec2BaseModelOutput`), since there are architecture-specific outputs besides the spectrogram, e.g. pitch and energy

anton-l reviewed (commented):
Hey @jaketae, great work on the port, looks very clean so far! 

For the next steps:
1) Upload the conversion script to `models/fastspeech2/` for future checkpoint ports :)
2) Try converting other checkpoints, e.g. one of each model size, one for each dataset (ljspeech, commonvoice en) and write integration tests for them, like you did for the first test. With this we'll make sure that there won't be any missing modules or config options for different model variants down the road, when we start mass-converting the leftover checkpoints.
3) Remove unsupported task-specific tests from `FastSpeech2ModelTester` and `FastSpeech2ModelTest`, e.g. `create_and_check_for_question_answering`. Then see if `FastSpeech2ModelTest()` is all green.

Commit:
feature: use custom modeling output cls

Commit:
fix: account for multi spk (common voice)

Commit:
feature: add conversion script

Commit:
style: run linter

Commit:
chore: fix wav2vec2 reference to fastspeech2

Commit:
fix: add fastspeech2tokenizer to auto

Commit:
chore: conform to convert func signature

Commit:
fix: robustify mean std configuration code

Commit:
Merge branch 'master' into add-fairseq-fs2

Commit:
chore: complete merge conflict

Commit:
chore: complete merge

Commit:
chore: adapt to new test dir structure

Commit:
chore: add `__init__.py` in tests dir

Commit:
chore: update docs, rm non existent params

Commit:
chore: `args` -> `config` var name change

Commit:
wip: start working on tests

Commit:
test: write integration tests

jaketae commented:
Hey @anton-l, here are some long overdue updates:

* Incorporate both LJSpeech and Common Voice models to the model hub
* Write unit/integration tests

At the moment, `test_initialization` is failing, and I'm still investigating how to fix it. Any input would greatly be appreciated. 

I've added a fairly simple `test_inference_integration` test to check that the expected mel-spectrogram output by the fairseq model is `torch.allclose` to the HF port. However, mel-spectrograms are too big in dimension, e.g., `(batch_size, seq_len, mel_dim)`, where `mel_dim` is typically 80. Therefore, I've sliced the first 100 elements instead. If you think there is a more robust check (i.e., more elements, different examples, etc), also please let me know!

Thank you.

Commit:
Merge branch 'main' into add-fairseq-fs2

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15773). All of your documentation changes will be reflected on that endpoint.

Commit:
docs: mv docs to en directory

Commit:
chore: make doc builder

jaketae committed:
Merge branch 'main' into add-fairseq-fs2

anton-l commented:
@LysandreJik @sgugger would be awesome if you take a look at this PR too, since it introduces a novel modality :)

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+]
+
+
+def lengths_to_padding_mask(lens: torch.LongTensor) -> torch.BoolTensor:
+    # from https://github.com/pytorch/fairseq/blob/main/fairseq/data/data_utils.py
+    bsz, max_lens = lens.size(0), torch.max(lens).item()
```
Please use proper names for variable. `lengths` is three character longer than `lens` and everyone will understand it. Likewise, use `max_lengths` and `batch_size` below.

jaketae replied:
Thanks for pointing that out, and apologies for making you repeat yourself in the comments above and below. I've gone through a first pass through the codebase to rename ambiguous/short variables (e.g., `x`), and I'll make sure to fix any remaining ones before the next review.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+                padding=(kernel_size - 1) // 2,
+            ),
+            nn.ReLU(),
+            nn.Conv1d(
+                hidden_dim,
+                in_dim,
+                kernel_size=kernel_size,
+                padding=(kernel_size - 1) // 2,
+            ),
```
```suggestion
            nn.Conv1d(hidden_dim, in_dim, kernel_size=kernel_size, padding=(kernel_size - 1) // 2),
```
Likewise.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+        self.dropout = self.dropout_module = nn.Dropout(dropout)
+
+    def forward(self, x):
+        # B x T x C
```
That comment is not very helpful, could you please put the names of the dimensions?

Plus we prefer using `hidden` to `x`.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+        self.ffn = PositionwiseFeedForward(embed_dim, hidden_dim, kernel_size, dropout=dropout)
+
+    def forward(self, x, padding_mask=None):
+        # B x T x C
```
Same here.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+
+class LengthRegulator(nn.Module):
+    def forward(self, x, durations):
+        # x: B x T x C
```
Same here and below. Better names would be great.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,700 @@
+
+    def forward(self, x):
+        # Input: B x T x C; Output: B x T
+        x = self.conv1(x.transpose(1, 2)).transpose(1, 2)
```
No reason to close this comment for now as it has not been addressed.

jaketae replied:
Apologies for the hasty resolve. I fixed in the latest commit; will keep this thread open for a reviewer to confirm!

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+        p_factor=1.0,
+        e_factor=1.0,
+    ):
+        # x: B x T x C
```
As before, please use the names of each dimension here.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+    def __init__(self, in_dim, n_channels, kernel_size, n_layers, dropout):
+        super(Postnet, self).__init__()
+        self.convolutions = nn.ModuleList()
+        assert kernel_size % 2 == 1
```
No assert in the model code base. Use a test and raise an appropriate error with a helpful message.

jaketae replied:
I'm removing the `assert` entirely becaue `FastSpeech2Config` already raises a `ValueError` if `kernel_size` is even. Does that sound okay?

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+        self.convolutions = nn.ModuleList()
+        assert kernel_size % 2 == 1
+        for i in range(n_layers):
+            cur_layers = (
+                [
+                    nn.Conv1d(
+                        in_dim if i == 0 else n_channels,
+                        n_channels if i < n_layers - 1 else in_dim,
+                        kernel_size=kernel_size,
+                        padding=((kernel_size - 1) // 2),
+                    ),
+                    nn.BatchNorm1d(n_channels if i < n_layers - 1 else in_dim),
+                ]
+                + ([nn.Tanh()] if i < n_layers - 1 else [])
+                + [nn.Dropout(dropout)]
+            )
```
This code is not very readable. After initializing the list with the two first objects, use a test to append `nn.Tanh` when necessary, then append `nn.Dropout(...)`

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+            self.convolutions.append(nn.Sequential(*cur_layers))
+
+    def forward(self, x):
+        x = x.transpose(1, 2)  # B x T x C -> B x C x T
```
The comment should go above, with proper names.

sgugger commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+        if self.embed_speaker is not None:
+            if speaker is None:
+                raise ValueError("`speaker` cannot be `None` for multi-speaker FastSpeech2.")
+            bsz, seq_len, _ = x.size()
+            emb = self.embed_speaker(speaker).expand(bsz, seq_len, -1)
+            x = self.spk_emb_proj(torch.cat([x, emb], dim=2))
+
+        x, out_lens, log_dur_out, pitch_out, energy_out = self.var_adaptor(
+            x, enc_padding_mask, durations, pitches, energies
+        )
```
Better names here and below please.

sgugger commented on src/transformers/models/fastspeech2/tokenization_fastspeech2.py:
```diff
@@ -0,0 +1,145 @@
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "fastspeech2": "https://huggingface.co/fastspeech2/resolve/main/vocab.txt",
```
Like before, we will need a real checkpoint here.

sgugger reviewed (commented):
Thanks for adding this new model! The code doesn't satisfy our [code style standard](https://huggingface.co/docs/transformers/main/en/add_new_model#code-style) so please make sure to update short variable names to something easier to understand.

Other than that, it all looks good to me!

jaketae committed:
Update docs/source/en/model_doc/fastspeech2.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/models/fastspeech2/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/models/auto/tokenization_auto.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/models/fastspeech2/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/models/fastspeech2/configuration_fastspeech2.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/models/fastspeech2/modeling_fastspeech2.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jaketae committed:
Update src/transformers/models/fastspeech2/modeling_fastspeech2.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Commit:
chore: cleanup layer logic, rename vars in `lengths_to_padding_mask`

jaketae commented on src/transformers/models/fastspeech2/configuration_fastspeech2.py:
```diff
@@ -0,0 +1,154 @@
+logger = logging.get_logger(__name__)
+
+FASTSPEECH2_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "fastspeech2": "https://huggingface.co/fastspeech2/resolve/main/config.json",
```
@anton-l, would we put this under `facebook/fastspeech2`?

jaketae replied:
Yes, will do!

Commit:
chore: rename variables to conform to hf code style

jaketae commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,698 @@
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    bsz, src_len = mask.size()
```
This was taken from https://github.com/huggingface/transformers/blob/69233cf03be5fbce0492f3997e139c4d05499e27/src/transformers/models/bart/modeling_bart.py#L106-L109.

I can certainly update the comments and variable names here. Would you also accept a small PR that renames them in `modeling_bart.py`?

Commit:
chore: cleanup imports, fix var names in lengths_to_padding_mask

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+        return out, out_lengths
+
+
+class VariancePredictor(nn.Module):
```
These classes should have a `FastSpeech2` prefix :-)

patrickvonplaten commented on src/transformers/models/fastspeech2/configuration_fastspeech2.py:
```diff
@@ -0,0 +1,154 @@
+        energy_max=3.2244551181793213,
+        energy_min=-4.9544901847839355,
+        initializer_range=0.0625,
+        mean=True,
```
not a huge fan of the naming here `mean` and `std` don't sound like flag variables here.

jaketae replied:
Yes, I will update them with `use_mean` and `use_standard_deviation`!

patrickvonplaten commented on src/transformers/models/fastspeech2/configuration_fastspeech2.py:
```diff
@@ -0,0 +1,154 @@
+        self.initializer_range = initializer_range
+        self.mean = mean
+        self.std = std
+        super().__init__(pad_token_id=pad_token_id, **kwargs)
```
can we move this above all the attributes? :-)

jaketae replied:
Yes, that makes sense. I'll move it right below the `ValueError` checks!

patrickvonplaten commented on src/transformers/models/fastspeech2/configuration_fastspeech2.py:
```diff
@@ -0,0 +1,154 @@
+        postnet_dropout=0.5,
+        vocab_size=75,
+        num_speakers=1,
+        var_pred_n_bins=256,
```
that's quite a lot of variables - We've made the mistake a bit with Wav2Vec2 that we left every parameter modifiable which led to configs becoming huge and a bit unreadable IMO e.g.:  https://huggingface.co/itaihay/wav2vec_asr_swbd/blob/main/config.json

Could we maybe in a first step, hardcode lots of those variables if they don't differ from checkpoint to checkpoint? This will allow us to have shorter, more concise configs. We can always go back and make them modifiable if needed . However we can't do it the other way ;-)

patrickvonplaten replied:
Parameters that we don't users to modify we could define as:

```
@property 
def var_pred_n_bins():
    return 256

jaketae replied:
That's a good point. I'll hard code them as `@property`s!

jaketae replied:
Hello @patrickvonplaten, I'm having a bit of trouble deciding on which parameters we want to allow users to modify. Currently, I have `var_pred_num_bins` and `mel_dim`. Do you think we should hard code other parameters too, such as `num_layers` or `kernel_size`?

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+
+
+@dataclass
+class FastSpeech2ModelOutput(ModelOutput):
```
We need some docs here

jaketae replied:
Added!

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+    energy: torch.FloatTensor = None
+
+
+# Copied from transformers.models.speech_to_text.modeling_speech_to_text.Speech2TextSinusoidalPositionalEmbedding with Speech2Text->FastSpeech2
```
nice!

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+class PositionwiseFeedForward(nn.Module):
+    def __init__(self, in_dim, hidden_dim, kernel_size, dropout):
+        super().__init__()
+        self.ffn = nn.Sequential(
```
Please avoid using `nn.Sequential` better to write it out

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+        return self.layer_norm(hidden + residual)
+
+
+class FFTLayer(nn.Module):
```
Can we add a `FastSpeech2` prefix to this class and all other ones?

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+class VariancePredictor(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.conv1 = nn.Sequential(
```
Please avoid `nn.Sequential`

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+            if i < n_layers - 1:
+                cur_layers.append(nn.Tanh())
+            cur_layers.append(nn.Dropout(dropout))
+            nn.init.xavier_uniform_(
```
init should happen in `_init_weights_` function

patrickvonplaten commented on src/transformers/models/fastspeech2/modeling_fastspeech2.py:
```diff
@@ -0,0 +1,697 @@
+        self.n_frames_per_step = config.n_frames_per_step
+        self.out_dim = config.output_frame_dim * config.n_frames_per_step
+
+        self.spk_emb_proj = None
```
Let's remove this no?

patrickvonplaten replied:
Why should we set a weight to None?

jaketae replied:
Multi-speaker checkpoints (e.g. common voice) will have a speaker embedding, but a single-speaker checkpoint like LJSpeech won't. Do you think there is perhaps a better way to deal with this?

patrickvonplaten commented on src/transformers/models/fastspeech2/tokenization_fastspeech2.py:
```diff
@@ -0,0 +1,145 @@
+
+class FastSpeech2Tokenizer(PreTrainedTokenizer):
+    """
+    Construct a FastSpeech2 tokenizer. Based on byte-level Byte-Pair-Encoding.
```
```suggestion
    Construct a FastSpeech2 tokenizer.
```

patrickvonplaten commented on src/transformers/models/fastspeech2/tokenization_fastspeech2.py:
```diff
@@ -0,0 +1,145 @@
+        text = text.strip()
+
+        # phonemize
+        if self.do_phonemize:
```
Can this ever be `False`?

jaketae replied:
I guess it could be if someone is training a model from scratch, but the pretrained checkpoints all require phonemization.

patrickvonplaten commented on tests/fastspeech2/test_modeling_fastspeech2.py:
```diff
@@ -0,0 +1,267 @@
+
+        # mel-spectrogram is too large (123, 80),
+        # so only check top-left 100 elements
+        expected_mel_spectrogram = torch.tensor(
```
How is the mel spectrogram transformed to actual audio?

patrickvonplaten replied:
Is there a most common "waveform-decoder" that we could add to the modeling file? Do those decoders have weights or not? Think the UserXP would be nicer, if the "waveform-decoder" is added to the modeling file as an additional head. If there are multiple ones, we could maybe just add the best one?

jaketae replied:
I agree that having `FastSpeech2ForWaveGeneration` would be quite nice to have. I think HiFi-GAN would be a sensible choice for a decoder since it's part of `fairseq`'s TTS pipeline. Some questions:

- Should HiFi-GAN code be part of `modeling_fastspeech2.py`?
- Do we still want to prefix HiFi-GAN modules with `FastSpeech2`?
- In the off chance that we want to use HiFi-GAN for other TTS models (e.g., [FastPitch](https://github.com/huggingface/transformers/issues/16349), which is next in line), would we be duplicating code?

cc @anton-l

patrickvonplaten replied:
I would answer all there questions above with "Yes" - happy to add all the code in the modeling file

jaketae replied:
Hey @patrickvonplaten, thanks for the feedback. I added HiFi-GAN in the lastest commits. Some questions:

1. **HiFi-GAN Trainability.** Should HiFi-GAN (or `FastSpeech2ForWaveformGeneration`) be trainable? HiFi-GAN is trained just like any normal model, but during inference, the user should call `remove_weight_norm`. For now, I just manually called the function in the `FastSpeech2ForWaveformGeneration.forward()` function, but this might mean that the class is not trainable. Alternatively, I could check if the model is in `eval` mode and only call `remove_weight_norm` if it's in `model.train` is `False`. Do you have a preference here?
2. **HiFi-GAN Configuration.** I hard-coded all HiFi-GAN parameters instead of polluting `FastSpeech2Config`. HiFi-GAN does come in 3 variations (check `config.jsons` [here](https://github.com/jik876/hifi-gan)), but the V1 model has the most amount of checkpoints, covering LJ Speech, VCTK, and a "universal" version for general-purpose decoding. The hard-coded parameters thus follow the V1 configuration.
3. **Waveform Output.** Because input tokens are padded, the output mel-spectrogram will also be of different lengths (although they will appear to be of equal lengths due to batching). When we pass this batch of padded mel-spectrograms to HiFi-GAN, the vocoder will also produce padded audio (again, due to batching, the waveforms may appear to be of equal length, but the shorter samples will contain meaningless noise at the end). I'm trying to think of a good UI/UX to deal with this. Any input would be greatly appreciated!

I realize this PR has been WIP for a very long time. Apologies for the delay, and thank you so much for the review!

patrickvonplaten commented on src/transformers/models/fastspeech2/configuration_fastspeech2.py:
````diff
@@ -0,0 +1,154 @@
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+    model_type = "fastspeech2"
````
@anton-l - do you think it's ok if the "general" `fastspeech2` name is taken here even thought the model is from MS originally? Fine for me personally, if Facebook's fastspeech2 is the "best" OS fastspeech implementation

Should we ask the MS authors if they are ok with this just to be sure it's fine?

jaketae replied:
Although I do think `fairseq`'s FastSpeech2 is likely the "best" OS implementation out there, I'm not too confident about the naming. I don't think we can ever put it under `microsoft/fastspeech2`; if anything, it should be `facebook/fastspeech2`, or maybe even `facebook/fairseq-fastspeech2`.

Has there ever been a case where we ported an "unofficial" model to the transformers library (although `fairseq` is a pretty reputable source)?

patrickvonplaten replied:
Think it's a first ;-) But fine to leave it as is here actually

patrickvonplaten reviewed (commented):
There are still some TODO's in this PR.

Mainly:
- add docstring everywhere
- replace `nn.Sequential` 

Now the big question is whether we should include the Waveform decoder in the modeling file or not. I'd personally be in favor of it, especially if it has trained weights. 
Note that we could also think about making use of a `FastSpeech2Processor` if necessary. Wdyt @anton-l @jaketae ?

jaketae committed:
Update src/transformers/models/fastspeech2/tokenization_fastspeech2.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

Commit:
fix: rename flag vars, use property for mel_dim, documentation

Commit:
fix: rm nn.sequential, add docstring, rename vars

Commit:
fix: update pos embed size with real checkpoint links

Commit:
style: run linter

Commit:
fix: rename vars, run linter

Commit:
fix: update to reflect no nn.sequential

Commit:
fix: account for ln -> layernorm

Commit:
fix: ln -> layernorm, rename var, update checkpoint url

Commit:
fix: add real checkpoint

Commit:
chore: add var_pred_num_bin to @property

Commit:
docs: specify input tensor shapes

Commit:
test: update test to reflect modeling code change

Commit:
feature: add hifi gan

Commit:
fix: rm hifigan config, hard code params

Commit:
fix: force apply remove_weight_norm, transpose for dim match

arampacha commented:
Hi, may I ask a couple of questions:
1. Is there some estimated time of merging for this PR?
2. Are there any experiments/examples of training using this implementation Fastspeech2?
And somewhat more general question:
3. Is there work in progress related to `text-to-speech` pipeline?

I'm working training/fine-tuning fastspeech2 models and would like to deploy using HF inference API.
Would be happy to help out with testing and contribute to pipeline and/or porting HiFiGAN or provide any assistance I can to speed things up with TTS

patrickvonplaten commented:
@anton-l, think it's not high priority but once you have a free day let's maybe add the model :-)

jaketae commented:
Wish I could have iterated faster to finish it earlier. Let me know if you need anything from me @anton-l @patrickvonplaten!

patrickvonplaten commented:
> Wish I could have iterated faster to finish it earlier. Let me know if you need anything from me @anton-l @patrickvonplaten!

@jaketae if you want, would you like to finish this PR maybe ? E.g. fix the failing tests and add **all** the necessary code (including the vocoders) to the modeling file?

patrickvonplaten commented:
@anton-l I still think we should go ahead with this PR - it's been stale for a while now. Wdyt?

anton-l commented:
@patrickvonplaten definitely! I'm trying to find some time alongside other tasks to work on FS2 :sweat_smile:

jaketae commented:
Hey @anton-l @patrickvonplaten, hope you're doing well! I was wondering where FS2 fits within the transformers road map, and whether it would make sense for me to update this branch and wrap up the work. I feel responsible for leaving this stale, and I still think it would be cool to have this feature pushed out. Thank you!

patrickvonplaten commented:
Hey @jaketae, 

We definitely still want to add the model - would you like to go ahead and update the branch?

patrickvonplaten commented:
Closing this for now as I don't think anybody will have time to work on it soon. 

If anybody from the community wants to give it a shot, please feel free to do so!

JuheonChu commented:
> I don't think anybody will have time to work on it soon. If anybody from the community wants to give it a shot, please feel free to do so!

Can I try working on this?

connor-henderson commented:
@JuheonChu I'm also interested in working on this, any interest in collaborating? Planning to get up to speed then open a fresh pr.

hollance commented:
Hi @JuheonChu @connor-henderson, internally at HF we've been discussing what to do with this model. This PR has a lot of good work done already and FastSpeech 2 is a popular model, so it makes a lot of sense to add this to Transformers. If you want to help with this, that's great. 😃 

There are a few different implementations of FastSpeech 2 out there and we'd like to add the best one to Transformers. As far as we can tell, there is a Conformer-based FastSpeech 2 model in ESPNet that outperforms regular FastSpeech 2, so that might be an interesting candidate. You can find the [Conformer FastSpeech 2 checkpoints here](https://huggingface.co/models?sort=downloads&search=espnet+fastspeech2+conformer).

In addition, PortaSpeech is very similar to FastSpeech 2 but it replaces the post-net and possibly a few other parts of the model. So that might be an interesting variation of FS2 to add as well.

We're still trying to figure out which of these models we'd prefer to have, but since you've shown an interest in working on this, I wanted to point out the direction we were thinking of taking these models internally. And I'd like to hear your opinions on this as well. 😄 

P.S. Make sure to check out the implementation of SpeechT5. It has many features in common with FastSpeech2 (such as the vocoder and the generation loop) and so we'd want to implement FS2 mostly in the same manner.

rajveer43 commented:
why anybody is not working on htis?
